tests/test_errors.py

```diff
@@ -1333,3 +1334,44 @@ def test_parse_error_on_creation(create_func: Any, arg: Any) -> None:
         GrammarParseError, match=re.escape("no viable alternative at input '${b'")
     ):
         create_func(arg)
+
+
+def test_cycle_when_iterating_over_parents() -> None:
+    c = OmegaConf.create({"x": {}})
+    x_node = c._get_node("x")
+    assert isinstance(x_node, DictConfig)
+    c._set_parent(x_node)
+    with pytest.raises(
+        OmegaConfBaseException,
+        match=re.escape("Cycle when iterating over parents of key `x`"),
+    ):
+        c._get_full_key("x")
+
+
+def test_get_full_key_failure_in_format_and_raise() -> None:
+    c = OmegaConf.create({"x": {}})
+    x_node = c._get_node("x")
+    assert isinstance(x_node, DictConfig)
+    # We create a cycle in the parent relationship that will trigger a RecursionError
+    # when trying to access `c.x`. This test verifies that this RecursionError is properly
+    # raised even if another exception occurs in `format_and_raise()` when trying to
+    # obtain the full key.
+    c._set_parent(x_node)
+
+    # For some reason in Python 3.6 the error message may vary depending on how this
+    # test is run (`nox -s coverage-3.6` vs `nox -s omegaconf-3.6`). As a result, we
+    # do not check the message.
+    if sys.version_info < (3, 7):
+        match = ""
+    else:
+        match = re.escape(
+            dedent(
+                """\
+                maximum recursion depth exceeded in comparison
+                    full_key: <unresolvable due to OmegaConfBaseException: Cycle when iterating over parents of key `x`
+                """
+            )
+        )
+
+    with pytest.raises(RecursionError, match=match):
+        c.x
```

Comment:
I am curious, where in that flow are we comparing objects (based on the error message, we are).

Reply:
Good question, I don't really understand why, since here the recursion is in `_get_flag()` and I see no comparison...
I thought it was just the standard recursion error message, but trying to repro on a simpler example, the "in comparison" doesn't show.
So I'm afraid I don't have a good answer for this. Not sure it's worth digging too deep into it...

Reply:
Maybe put a breakpoint in the `___eq__` override and see if you hit it?

Reply:
So actually it's because of the `assert isinstance(parent, Container)` in `Node._get_parent()`. It's a bit messed up (as far as I can tell the `isinstance()` call is *not* part of the recursion stack trace, except when Python throws), but that's how it is right now...
If this breaks in the future I'll just handcraft a different error for the purpose of the test.

Reply:
I don't follow:
Why does `assert parent is None or isinstance(parent, Container)` trigger an equality check?

Reply:
There's no equality check, my (admittedly superficial) understanding is that `isinstance()` is considered a comparison, and somehow when the recursion error triggers it is linked to this check. But it seems to be an implementation detail mostly -- messing around with the code I was able to get it to have / not have the "in comparison" part of the error just with irrelevant print statements.

I changed the test in f387495 to just not test the specific exeption message coming from Python, since it doesn't seem reliable.

Reply:
Okay, I don't think instancecheck is comparing the object.
it should be an operating on the type of the object.

Reply:
> Okay, I don't think instancecheck is comparing the object.

It doesn't -- at least not in the sense of calling `__eq__()`. However, for the purpose of Python's max recursion error message, it seems to be considered a comparison.